File: src/reviewiq/ci.py
from __future__ import annotations

"""
ReviewIQ CI Mode — GitHub Actions integration.

Handles webhook events (PR opened, push, comment) using the shared engine.
Posts results as PR comments and persists state to hidden PR comments.
"""

import json
import os
import re
import sys
import urllib.request

from reviewiq import state as st
from reviewiq import git
from reviewiq import engine


def log(msg: str) -> None:
    print(f"[reviewiq:ci] {msg}", file=sys.stderr)


def gh_api(method: str, endpoint: str, data: dict | None = None) -> dict | list | None:
    """Call GitHub API."""
    url = f"https://api.github.com{endpoint}"
    body = json.dumps(data).encode() if data else None
    headers = {
        "Authorization": f"token {os.environ['GITHUB_TOKEN']}",
        "Accept": "application/vnd.github.v3+json",
    }
    req = urllib.request.Request(url, data=body, method=method, headers=headers)
    if body:
        req.add_header("Content-Type", "application/json")
    with urllib.request.urlopen(req) as resp:
        if resp.status == 204:
            return None
        return json.loads(resp.read().decode())


def post_comment(repo: str, pr_number: int, body: str) -> None:
    """Post a visible comment on the PR."""
    gh_api("POST", f"/repos/{repo}/issues/{pr_number}/comments", {"body": body})


def get_pr_metadata(repo: str, pr_number: int) -> dict:
    """Fetch PR metadata from GitHub API."""
    pr_data = gh_api("GET", f"/repos/{repo}/pulls/{pr_number}")
    return {
        "title": pr_data["title"],
        "body": pr_data.get("body") or "",
        "author": pr_data["user"]["login"],
        "base_branch": pr_data["base"]["ref"],
        "head_branch": pr_data["head"]["ref"],
        "base_sha": pr_data["base"]["sha"],
        "head_sha": pr_data["head"]["sha"],
    }


def run(repo: str, pr_number: int, event_type: str, comment_body: str = "") -> None:
    """Main CI entry point."""
    log(f"Event: {event_type}, PR: #{pr_number}, Repo: {repo}")

    review_state = st.load(pr_number, repo, prefer="auto")
    log(f"State loaded: {review_state['summary']['total_findings']} findings, "
        f"{len(review_state['review_rounds'])} rounds")

    if event_type == "opened":
        _handle_opened(review_state, repo, pr_number)
    elif event_type == "synchronize":
        _handle_synchronize(review_state, repo, pr_number)
    elif event_type == "comment":
        if re.search(r"@review-agent", comment_body, re.IGNORECASE):
            _handle_comment(review_state, repo, pr_number, comment_body)
        else:
            log("Comment doesn't mention @review-agent, skipping")
    else:
        log(f"Unknown event type: {event_type}")
        sys.exit(1)


def _handle_opened(review_state: dict, repo: str, pr_number: int) -> None:
    log("Handling PR opened")

    meta = get_pr_metadata(repo, pr_number)
    review_state["pr"].update({
        "title": meta["title"],
        "author": meta["author"],
        "base_branch": meta["base_branch"],
        "head_branch": meta["head_branch"],
    })

    diff = git.get_diff(meta["base_sha"], meta["head_sha"])
    changed_files = git.get_changed_files(meta["base_sha"], meta["head_sha"])
    file_contents = git.read_files(changed_files)
    history = git.get_file_history(changed_files)

    response = engine.run_review(
        review_state,
        diff=diff,
        file_contents=file_contents,
        history=history,
        changed_files=changed_files,
        head_sha=meta["head_sha"],
        base_sha=meta["base_sha"],
        pr_title=meta["title"],
        pr_author=meta["author"],
        pr_body=meta["body"],
        base_branch=meta["base_branch"],
        head_branch=meta["head_branch"],
    )

    st.save(review_state, targets="both")
    post_comment(repo, pr_number, response)
    log(f"Review posted. {review_state['summary']['total_findings']} findings tracked.")


def _handle_synchronize(review_state: dict, repo: str, pr_number: int) -> None:
    log("Handling push (incremental re-review)")

    meta = get_pr_metadata(repo, pr_number)
    diff = git.get_diff(meta["base_sha"], meta["head_sha"])
    changed_files = git.get_changed_files(meta["base_sha"], meta["head_sha"])
    file_contents = git.read_files(changed_files)
    incremental_diff = git.get_incremental_diff(review_state, meta["head_sha"])

    response = engine.run_check(
        review_state,
        diff=diff,
        file_contents=file_contents,
        changed_files=changed_files,
        head_sha=meta["head_sha"],
        base_sha=meta["base_sha"],
        incremental_diff=incremental_diff,
        pr_title=meta["title"],
        pr_author=meta["author"],
        base_branch=meta["base_branch"],
        head_branch=meta["head_branch"],
    )

    st.save(review_state, targets="both")
    post_comment(repo, pr_number, response)
    log(f"Re-review posted. Open: {review_state['summary']['open']}, Resolved: {review_state['summary']['resolved']}")


def _handle_comment(review_state: dict, repo: str, pr_number: int, comment_body: str) -> None:
    command = re.sub(r"@review-agent\s*", "", comment_body, flags=re.IGNORECASE).strip()
    log(f"Handling comment: {command[:80]}...")

    meta = get_pr_metadata(repo, pr_number)
    changed_files = git.get_changed_files(meta["base_sha"], meta["head_sha"])
    file_contents = git.read_files(changed_files)

    finding_id = None
    finding_match = re.search(r"(?:finding|#)\s*(\d+)", command, re.IGNORECASE)
    if finding_match:
        finding_id = int(finding_match.group(1))

    response = engine.run_ask(review_state, command, file_contents, finding_id=finding_id)

    st.save(review_state, targets="both")
    post_comment(repo, pr_number, response)
    log("Reply posted")

File: internal/ci/ci.go
// Package ci handles GitHub Actions webhook events.
package ci

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/Sanmanchekar/reviewiq/internal/engine"
	gitops "github.com/Sanmanchekar/reviewiq/internal/git"
	gh "github.com/Sanmanchekar/reviewiq/internal/github"
	"github.com/Sanmanchekar/reviewiq/internal/state"
)

func log(msg string) { fmt.Fprintf(os.Stderr, "[reviewiq:ci] %s\n", msg) }

func Run(repo string, prNumber int, eventType, commentBody string) {
	log(fmt.Sprintf("Event: %s, PR: #%d, Repo: %s", eventType, prNumber, repo))

	s := state.Load(prNumber, repo)
	s.PR.Repo = repo
	log(fmt.Sprintf("State loaded: %d findings, %d rounds", s.Summary.TotalFindings, len(s.ReviewRounds)))

	owner, repoName := splitRepo(repo)

	switch eventType {
	case "opened":
		handleOpened(s, owner, repoName, repo, prNumber)
	case "synchronize":
		handleSynchronize(s, owner, repoName, repo, prNumber)
	case "comment":
		handleComment(s, owner, repoName, repo, prNumber, commentBody)
	default:
		log("Unknown event type: " + eventType)
		os.Exit(1)
	}
}

func splitRepo(repo string) (string, string) {
	parts := strings.SplitN(repo, "/", 2)
	if len(parts) != 2 {
		return "", repo
	}
	return parts[0], parts[1]
}

func handleOpened(s *state.ReviewState, owner, repoName, fullRepo string, prNumber int) {
	log("Handling PR opened — full review")
	prInfo, err := gh.GetPR(owner, repoName, prNumber)
	if err != nil {
		log("Failed to get PR: " + err.Error())
		return
	}
	s.PR.Title = prInfo.Title
	s.PR.Author = prInfo.Author
	s.PR.BaseBranch = prInfo.BaseBranch
	s.PR.HeadBranch = prInfo.HeadBranch

	diff := gitops.GetDiff(prInfo.BaseSHA, prInfo.HeadSHA)
	changedFiles := gitops.GetChangedFiles(prInfo.BaseSHA, prInfo.HeadSHA)
	fileContents := gitops.ReadFiles(changedFiles)
	history := gitops.GetFileHistory(changedFiles)

	response, err := engine.RunReview(s, diff, fileContents, history, changedFiles,
		prInfo.HeadSHA, prInfo.BaseSHA, prInfo.Title, prInfo.Author, prInfo.Body, prInfo.BaseBranch, prInfo.HeadBranch)
	if err != nil {
		log("Review failed: " + err.Error())
		return
	}

	// Post inline comments for each finding
	round := len(s.ReviewRounds)
	var comments []gh.InlineComment
	for _, f := range s.Findings {
		if f.CreatedRound == round {
			comments = append(comments, gh.InlineComment{
				Path: f.File,
				Line: f.Line,
				Body: gh.FormatSuggestion(f.Severity, f.Title, f.Problem, f.SuggestedFix, f.FixRationale),
			})
		}
	}

	if len(comments) > 0 {
		event := "COMMENT"
		if s.Summary.Open > 0 {
			event = "REQUEST_CHANGES"
		}
		summary := fmt.Sprintf("## ReviewIQ Review — Round %d\n\n**%d findings** | Assessment: **%s**",
			round, s.Summary.TotalFindings, s.Summary.Assessment)
		if err := gh.PostReview(owner, repoName, prNumber, prInfo.HeadSHA, summary, event, comments); err != nil {
			log("Failed to post review with inline comments: " + err.Error())
			// Fallback to plain comment
			_ = gh.PostPRComment(owner, repoName, prNumber, response)
		}
	} else {
		_ = gh.PostPRComment(owner, repoName, prNumber, response)
	}

	state.Save(s)
	log(fmt.Sprintf("Review posted. %d findings, %d inline comments.", s.Summary.TotalFindings, len(comments)))
}

func handleSynchronize(s *state.ReviewState, owner, repoName, fullRepo string, prNumber int) {
	log("Handling push — incremental re-review")
	prInfo, err := gh.GetPR(owner, repoName, prNumber)
	if err != nil {
		log("Failed to get PR: " + err.Error())
		return
	}

	diff := gitops.GetDiff(prInfo.BaseSHA, prInfo.HeadSHA)
	changedFiles := gitops.GetChangedFiles(prInfo.BaseSHA, prInfo.HeadSHA)
	fileContents := gitops.ReadFiles(changedFiles)
	incDiff := gitops.GetIncrementalDiff(s, prInfo.HeadSHA)

	response, err := engine.RunCheck(s, diff, fileContents, changedFiles,
		prInfo.HeadSHA, prInfo.BaseSHA, incDiff, prInfo.Title, prInfo.Author, prInfo.BaseBranch, prInfo.HeadBranch)
	if err != nil {
		log("Re-review failed: " + err.Error())
		return
	}

	// Post inline comments only for NEW findings in this round
	round := len(s.ReviewRounds)
	var comments []gh.InlineComment
	for _, f := range s.Findings {
		if f.CreatedRound == round {
			comments = append(comments, gh.InlineComment{
				Path: f.File,
				Line: f.Line,
				Body: gh.FormatSuggestion(f.Severity, f.Title, f.Problem, f.SuggestedFix, f.FixRationale),
			})
		}
	}

	if len(comments) > 0 {
		summary := fmt.Sprintf("## ReviewIQ Re-review — Round %d\n\nOpen: %d | Resolved: %d | New: %d | Assessment: **%s**",
			round, s.Summary.Open, s.Summary.Resolved, len(comments), s.Summary.Assessment)
		if err := gh.PostReview(owner, repoName, prNumber, prInfo.HeadSHA, summary, "COMMENT", comments); err != nil {
			log("Failed to post inline comments: " + err.Error())
			_ = gh.PostPRComment(owner, repoName, prNumber, response)
		}
	} else {
		// No new findings — post summary as comment
		_ = gh.PostPRComment(owner, repoName, prNumber, response)
	}

	state.Save(s)
	log(fmt.Sprintf("Re-review posted. Open: %d, Resolved: %d", s.Summary.Open, s.Summary.Resolved))
}

func handleComment(s *state.ReviewState, owner, repoName, fullRepo string, prNumber int, commentBody string) {
	if !regexp.MustCompile(`(?i)@review-agent`).MatchString(commentBody) {
		log("Comment doesn't mention @review-agent, skipping")
		return
	}

	command := regexp.MustCompile(`(?i)@review-agent\s*`).ReplaceAllString(commentBody, "")
	command = strings.TrimSpace(command)
	log(fmt.Sprintf("Handling comment command: %.80s", command))

	// Detect structured commands
	lowerCmd := strings.ToLower(command)

	switch {
	case strings.HasPrefix(lowerCmd, "resolve") || strings.HasPrefix(lowerCmd, "approve"):
		handleResolveComment(s, owner, repoName, fullRepo, prNumber)
	case strings.HasPrefix(lowerCmd, "recheck") || strings.HasPrefix(lowerCmd, "check"):
		handleSynchronize(s, owner, repoName, fullRepo, prNumber)
	default:
		// Freeform ask/explain
		handleAskComment(s, owner, repoName, fullRepo, prNumber, command)
	}
}

func handleResolveComment(s *state.ReviewState, owner, repoName, fullRepo string, prNumber int) {
	log("Handling resolve command")

	open := s.OpenFindings()
	if len(open) == 0 {
		msg := "## ReviewIQ Resolve\n\nNo open findings. PR is already clear."
		_ = gh.PostPRComment(owner, repoName, prNumber, msg)
		// Approve
		_ = gh.PostReview(owner, repoName, prNumber, s.Summary.LastReviewedSHA,
			"All findings resolved — ReviewIQ", "APPROVE", nil)
		s.Summary.Assessment = "APPROVE"
		state.Save(s)
		return
	}

	// Check each finding against current code
	round := len(s.ReviewRounds) + 1
	resolvedCount := 0
	for _, f := range open {
		// Read current file to check if fix was applied
		content, err := gh.GetFileContent(owner, repoName, s.Summary.LastReviewedSHA, f.File)
		if err != nil {
			continue
		}
		// If the suggested fix text appears in the file, consider it resolved
		if f.SuggestedFix != "" && strings.Contains(content, strings.TrimSpace(f.SuggestedFix)) {
			_ = s.TransitionFinding(f.ID, "resolved", "Fix confirmed in code", round)
			resolvedCount++
		}
	}

	remaining := s.OpenFindings()
	var msg string
	if len(remaining) == 0 {
		msg = fmt.Sprintf("## ReviewIQ Resolve — Round %d\n\nAll %d findings resolved. PR APPROVED.", round, resolvedCount)
		_ = gh.PostReview(owner, repoName, prNumber, s.Summary.LastReviewedSHA,
			"All findings resolved — ReviewIQ", "APPROVE", nil)
		s.Summary.Assessment = "APPROVE"
	} else {
		var lines []string
		for _, f := range remaining {
			lines = append(lines, fmt.Sprintf("- **Finding %d** [%s]: %s — `%s:%d`", f.ID, f.Severity, f.Title, f.File, f.Line))
		}
		msg = fmt.Sprintf("## ReviewIQ Resolve — Round %d\n\nResolved: %d | Still open: %d\n\n%s\n\nPush fixes and comment `@review-agent resolve` again.",
			round, resolvedCount, len(remaining), strings.Join(lines, "\n"))
	}

	_ = gh.PostPRComment(owner, repoName, prNumber, msg)
	state.Save(s)
	log(fmt.Sprintf("Resolve: %d resolved, %d remaining", resolvedCount, len(remaining)))
}

func handleAskComment(s *state.ReviewState, owner, repoName, fullRepo string, prNumber int, command string) {
	prInfo, err := gh.GetPR(owner, repoName, prNumber)
	if err != nil {
		log("Failed to get PR metadata: " + err.Error())
		return
	}
	changedFiles := gitops.GetChangedFiles(prInfo.BaseSHA, prInfo.HeadSHA)
	fileContents := gitops.ReadFiles(changedFiles)

	findingID := 0
	if m := regexp.MustCompile(`(?i)(?:finding|#)\s*(\d+)`).FindStringSubmatch(command); len(m) > 1 {
		fmt.Sscanf(m[1], "%d", &findingID)
	}

	response, err := engine.RunAsk(s, command, fileContents, findingID, changedFiles)
	if err != nil {
		log("Ask failed: " + err.Error())
		return
	}
	state.Save(s)
	_ = gh.PostPRComment(owner, repoName, prNumber, response)
	log("Reply posted")
}
